### PR TITLE
Fix install stream snapshots on graceful shutdown

### PR DESCRIPTION
**Pull request by MauriceVanVeen**
Upon graceful shutdown, stream snapshots aren't installed:
stream.go
```go
} else {
	// Always attempt snapshot on clean exit.
	n.InstallSnapshot(mset.stateSnapshotLocked())
	n.Stop()
}
```

But raft.go exits immediately since it's in Closed state
```go
if n.State() == Closed {
        return errNodeClosed
}
```

This turns out to be because of calling either of `close(mset.mqch)` or `close(mset.qch)`. Which then calls the following upon leaving `monitorStream`:
```go
	// Make sure to stop the raft group on exit to prevent accidental memory bloat.
	// This should be below the checkInMonitor call though to avoid stopping it out
	// from underneath the one that is running since it will be the same raft node.
	defer n.Stop()
```

This PR proposes to skip running `n.Stop()` from `monitorStream` when we know we're closing / have called `mset.stop()`. Which will already take care of calling either `n.Stop()` or `n.Delete()` so there's no need in stopping from `monitorStream` in that case.

TLDR; this ensures stream snapshots are installed upon shutdown.

Signed-off-by: Maurice van Veen <github@mauricevanveen.com>

### CODE REVIEW

**Comment by wallyqs** [comment]
Need to fix this one:
```
=== RUN   TestJetStreamClusterStreamResetPreacks
    jetstream_cluster_3_test.go:6071: Not correct state yet: {Msgs:0 Bytes:0 FirstSeq:0 FirstTime:0001-01-01 00:00:00 +0000 UTC LastSeq:0 LastTime:0001-01-01 00:00:00 +0000 UTC NumSubjects:0 Subjects:map[] NumDeleted:0 Deleted:[] Lost:<nil> Consumers:1}
--- FAIL: TestJetStreamClusterStreamResetPreacks (13.54s)
```

File: server/jetstream_cluster.go [hunk 1]
```diff
@@ -2263,9 +2263,11 @@ func (js *jetStream) monitorStream(mset *stream, sa *streamAssignment, sendSnaps
 	// from underneath the one that is running since it will be the same raft node.
 	defer func() {
 		// We might be closing during shutdown, don't pre-emptively stop here since we'll still want to install snapshots.
+		mset.mu.RLock()
 		if !mset.closed.Load() {
```
Comment (derekcollison) [inline server/jetstream_cluster.go:2267]: This is atomic, so don't need lock for that and since we already have `n` dereferenced not sure why these are needed?

File: server/jetstream_cluster.go [hunk 2]
```diff
@@ -3140,13 +3140,10 @@ func (js *jetStream) applyStreamEntries(mset *stream, ce *CommittedEntry, isReco
 				}
 			}
 
-			if !isRecovering && !mset.IsLeader() {
+			if isRecovering || !mset.IsLeader() {
 				if err := mset.processSnapshot(ss); err != nil {
```
Comment (derekcollison) [inline server/jetstream_cluster.go:3144]: This could be dangerous I think since on recover we could think we need to delete legit data..

File: server/jetstream_cluster.go (hunk 1 above)
Comment (MauriceVanVeen) [reply server/jetstream_cluster.go:2266]: Removed it.

File: server/jetstream_cluster.go (hunk 2 above)
Comment (derekcollison) [reply server/jetstream_cluster.go:3142]: Looked a bit closer, could be ok.

**Review by derekcollison — APPROVED** [review]
LGTM
